sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -298,6 +300,30 @@ class Analyzer(override val catalogManager: CatalogManager)
       CleanupAliases)
   )
 
+  /**
+   * Initialize the tag of Cast so we can distinguish if Cast is auto-generated by Analyzer and
+   * strip auto-generated cast when use Cast.sql.
+   */
+  object InitializeCastTag extends Rule[LogicalPlan] {
+    private def setCastTag(plan: LogicalPlan): Unit = {
+      plan.resolveOperators { case p =>
+        p.transformExpressions {
+          case cast: CastBase =>
+            cast.setTagValue(Cast.AUTO_GENERATED_TAG, false)
```

Comment:
I think a more explicit and easy way to find user-specified casts is:
1. the cast created in the SQL parser
2. the cast created in `Column.cast`

This rule catches all casts that before the plan is analyzed, but it's possible that spark may re-analyze a query plan (with dataframe API).

Reply:
I thought about this but then chose the current way ..

Reply:
@cloud-fan  Do we need care about dsl.cast ? e.g. `"c".attr.cast()`

Reply:
that's for test only, but from the test writer's point of view, it's user-specified as well, and let's handle it too.